src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.WaitThread.cs

```diff
@@ -0,0 +1,444 @@
+            /// Otherwise, the unregistration of the wait handle is queued on the wait thread.
+            /// </remarks>
+            public void UnregisterWait(RegisteredWaitHandle handle)
+            {
+                UnregisterWait(handle, handle.IsBlocking);
+            }
+
+            /// <summary>
+            /// Unregister a wait handle.
+            /// </summary>
+            /// <param name="handle">The wait handle to unregister.</param>
+            /// <param name="blocking">Should the unregistration block on the full unregistration and all pending callback completion.</param>
+            private void UnregisterWait(RegisteredWaitHandle handle, bool blocking)
+            {
+                bool pendingUnregistration = false;
+                // TODO: Optimization: Try to unregister wait directly if it isn't being waited on.
+                _removesLock.Acquire();
+                try
+                {
+                    // If this handle is not already pending removal and hasn't already been removed
+                    if (Array.IndexOf(_registeredWaits, handle) != -1 && Array.IndexOf(_pendingRemoves, handle) == -1)
+                    {
+                        _pendingRemoves[_numPendingRemoves++] = handle;
+                        _changeHandlesEvent.Set(); // Tell the wait thread that there are changes pending.
+                        pendingUnregistration = true;
+                    }
+                }
+                finally
+                {
+                    _removesLock.Release();
+                }
+
+                // We do not need to wait for the Wait Thread to actually remove the handle from its array as long as
+                // the wait subsystem's implementation of WaitAny (CoreRT's/Win32's) observes set wait handles
+                // in the order of the array and in order of being set. This is true in both cases as far as we can tell.
+                // Below are a few cases where this is important:
+                // * Wait thread is between processing removals and WaitAny or still processing removals when this method is called.
+                //   The thread that called this method then disposes its wait handle
+                //   - Wait thread needs to not sleep from _changeHandlesEvent already being set.
+                // * Wait thread is in WaitAny. User thread calls Unregister and then sets the relevant event, both before the Wait thread wakes up
+                //   - Wait thread needs to observe that this event was unregistered before it was signaled.
+                // * Wait thread is in WaitAny. User thread sets the relevant event and then calls Unregister, both before the Wait thread wakes up
+                //   - Wait thread needs to observe that this event was signaled before it was unregistered
+
+                if (pendingUnregistration && blocking)
```

Comment:
This is fixed and now tested by the `BlockingUnregisterBlocksEvenIfCallbackExecuting` test.